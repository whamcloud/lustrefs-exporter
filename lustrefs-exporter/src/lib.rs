// Copyright (c) 2024 DDN. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

pub mod brw_stats;
pub mod host;
pub mod jobstats;
pub mod llite;
pub mod lnet;
pub mod openmetrics;
pub mod quota;
pub mod service;
pub mod stats;

use axum::{
    http::{self, StatusCode},
    response::{IntoResponse, Response},
};
use lustre_collector::{LustreCollectorError, TargetVariant};

#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error(transparent)]
    Http(#[from] http::Error),
    #[error(transparent)]
    TaskJoin(#[from] tokio::task::JoinError),
    #[error(transparent)]
    Io(#[from] std::io::Error),
    #[error(transparent)]
    LustreCollector(#[from] LustreCollectorError),
    #[error(transparent)]
    Utf8(#[from] std::str::Utf8Error),
    #[error("Could not find match for {0} in {1}")]
    NoCap(&'static str, String),
    #[error(transparent)]
    Otel(#[from] opentelemetry_sdk::metrics::MetricError),
}

impl IntoResponse for Error {
    fn into_response(self) -> Response {
        tracing::warn!("{self}");

        StatusCode::INTERNAL_SERVER_ERROR.into_response()
    }
}

trait LabelProm {
    fn to_prom_label(&self) -> &'static str;
}

impl LabelProm for TargetVariant {
    fn to_prom_label(&self) -> &'static str {
        match self {
            TargetVariant::Ost => "ost",
            TargetVariant::Mgt => "mgt",
            TargetVariant::Mdt => "mdt",
        }
    }
}
