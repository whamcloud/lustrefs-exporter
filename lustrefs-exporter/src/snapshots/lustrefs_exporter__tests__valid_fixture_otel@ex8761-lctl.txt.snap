---
source: lustrefs-exporter/src/lib.rs
expression: x
input_file: lustre-collector/src/fixtures/valid/ex8761-lctl.txt
---
# HELP lustre_health_healthy Indicates whether the Lustre server is healthy or not. 1 is healthy, 0 is unhealthy.
# TYPE lustre_health_healthy gauge
lustre_health_healthy{} 0
# HELP lustre_lnet_mem_used Gives information about Lustre LNet memory usage.
# TYPE lustre_lnet_mem_used gauge
lustre_lnet_mem_used{} 66923896
# HELP lustre_mem_used Gives information about Lustre memory usage.
# TYPE lustre_mem_used gauge
lustre_mem_used{} 242840621
# HELP lustre_mem_used_max Gives information about Lustre maximum memory usage.
# TYPE lustre_mem_used_max counter
lustre_mem_used_max{} 244436941
# HELP lustre_ldlm_canceld_stats Gives information about LDLM Canceld service.
# TYPE lustre_ldlm_canceld_stats counter
lustre_ldlm_canceld_stats{operation="ldlm_cancel"} 24
lustre_ldlm_canceld_stats{operation="req_active"} 24
lustre_ldlm_canceld_stats{operation="req_qdepth"} 24
lustre_ldlm_canceld_stats{operation="req_timeout"} 24
lustre_ldlm_canceld_stats{operation="req_waittime"} 24
lustre_ldlm_canceld_stats{operation="reqbuf_avail"} 50
# HELP lustre_ldlm_cbd_stats Gives information about LDLM Callback service.
# TYPE lustre_ldlm_cbd_stats counter
lustre_ldlm_cbd_stats{operation="ldlm_bl_callback"} 4
lustre_ldlm_cbd_stats{operation="req_active"} 4
lustre_ldlm_cbd_stats{operation="req_qdepth"} 4
lustre_ldlm_cbd_stats{operation="req_timeout"} 4
lustre_ldlm_cbd_stats{operation="req_waittime"} 4
lustre_ldlm_cbd_stats{operation="reqbuf_avail"} 9
# HELP lustre_inodes_free The number of inodes (objects) available.
# TYPE lustre_inodes_free gauge
lustre_inodes_free{component="mdt",target="fs-MDT0000"} 1885252
lustre_inodes_free{component="mgt",target="MGS"} 32555
lustre_inodes_free{component="ost",target="fs-OST0000"} 39844
lustre_inodes_free{component="ost",target="fs-OST0001"} 39876
# HELP lustre_inodes_maximum The maximum number of inodes (objects) the filesystem can hold.
# TYPE lustre_inodes_maximum gauge
lustre_inodes_maximum{component="mdt",target="fs-MDT0000"} 1885696
lustre_inodes_maximum{component="mgt",target="MGS"} 32768
lustre_inodes_maximum{component="ost",target="fs-OST0000"} 40960
lustre_inodes_maximum{component="ost",target="fs-OST0001"} 40960
# HELP lustre_available_kilobytes Number of kilobytes readily available in the pool.
# TYPE lustre_available_kilobytes gauge
lustre_available_kilobytes{component="mdt",target="fs-MDT0000"} 2365496
lustre_available_kilobytes{component="mgt",target="MGS"} 463060
lustre_available_kilobytes{component="ost",target="fs-OST0000"} 4037172
lustre_available_kilobytes{component="ost",target="fs-OST0001"} 4037172
# HELP lustre_free_kilobytes Number of kilobytes allocated to the pool.
# TYPE lustre_free_kilobytes gauge
lustre_free_kilobytes{component="mdt",target="fs-MDT0000"} 2599604
lustre_free_kilobytes{component="mgt",target="MGS"} 489272
lustre_free_kilobytes{component="ost",target="fs-OST0000"} 4105984
lustre_free_kilobytes{component="ost",target="fs-OST0001"} 4105984
# HELP lustre_capacity_kilobytes Capacity of the pool in kilobytes.
# TYPE lustre_capacity_kilobytes gauge
lustre_capacity_kilobytes{component="mdt",target="fs-MDT0000"} 2602832
lustre_capacity_kilobytes{component="mgt",target="MGS"} 491092
lustre_capacity_kilobytes{component="ost",target="fs-OST0000"} 4108388
lustre_capacity_kilobytes{component="ost",target="fs-OST0001"} 4108388
# HELP lustre_exports_total Total number of times the pool has been exported.
# TYPE lustre_exports_total gauge
lustre_exports_total{component="mdt",target="fs-MDT0000"} 31
lustre_exports_total{component="mgt",target="MGS"} 8
lustre_exports_total{component="ost",target="fs-OST0000"} 8
lustre_exports_total{component="ost",target="fs-OST0001"} 8
# HELP lustre_exports_dirty_total Total number of exports that have been marked dirty.
# TYPE lustre_exports_dirty_total gauge
lustre_exports_dirty_total{component="ost",target="fs-OST0000"} 0
lustre_exports_dirty_total{component="ost",target="fs-OST0001"} 0
# HELP lustre_exports_granted_total Total number of exports that have been marked granted.
# TYPE lustre_exports_granted_total gauge
lustre_exports_granted_total{component="ost",target="fs-OST0000"} 272832
lustre_exports_granted_total{component="ost",target="fs-OST0001"} 272832
# HELP lustre_exports_pending_total Total number of exports that have been marked pending.
# TYPE lustre_exports_pending_total gauge
lustre_exports_pending_total{component="ost",target="fs-OST0000"} 0
lustre_exports_pending_total{component="ost",target="fs-OST0001"} 0
# HELP lustre_lock_contended_total Number of contended locks.
# TYPE lustre_lock_contended_total gauge
lustre_lock_contended_total{component="mdt",target="fs-MDT0000"} 32
lustre_lock_contended_total{component="ost",target="fs-OST0000"} 32
lustre_lock_contended_total{component="ost",target="fs-OST0001"} 32
# HELP lustre_lock_contention_seconds_total Time in seconds during which locks were contended.
# TYPE lustre_lock_contention_seconds_total gauge
lustre_lock_contention_seconds_total{component="mdt",target="fs-MDT0000"} 2
lustre_lock_contention_seconds_total{component="ost",target="fs-OST0000"} 2
lustre_lock_contention_seconds_total{component="ost",target="fs-OST0001"} 2
# HELP lustre_connected_clients Number of connected clients.
# TYPE lustre_connected_clients gauge
lustre_connected_clients{component="mdt",target="fs-MDT0000"} 0
# HELP lustre_lock_count_total Number of locks.
# TYPE lustre_lock_count_total gauge
lustre_lock_count_total{component="mdt",target="fs-MDT0000"} 96
lustre_lock_count_total{component="ost",target="fs-OST0000"} 0
lustre_lock_count_total{component="ost",target="fs-OST0001"} 0
# HELP lustre_lock_timeout_total Number of lock timeouts.
# TYPE lustre_lock_timeout_total counter
lustre_lock_timeout_total{component="mdt",target="fs-MDT0000"} 0
lustre_lock_timeout_total{component="ost",target="fs-OST0000"} 0
lustre_lock_timeout_total{component="ost",target="fs-OST0001"} 0
# HELP lustre_oss_ost_stats OSS ost stats.
# TYPE lustre_oss_ost_stats gauge
lustre_oss_ost_stats{operation="obd_ping",units="usecs"} 27
lustre_oss_ost_stats{operation="ost_connect",units="usecs"} 27
lustre_oss_ost_stats{operation="ost_create",units="usecs"} 32
lustre_oss_ost_stats{operation="ost_get_info",units="usecs"} 16
lustre_oss_ost_stats{operation="req_active",units="reqs"} 102
lustre_oss_ost_stats{operation="req_qdepth",units="reqs"} 102
lustre_oss_ost_stats{operation="req_timeout",units="secs"} 102
lustre_oss_ost_stats{operation="req_waittime",units="usecs"} 102
lustre_oss_ost_stats{operation="reqbuf_avail",units="bufs"} 207
# HELP lustre_oss_ost_create_stats OSS ost_create stats.
# TYPE lustre_oss_ost_create_stats gauge
lustre_oss_ost_create_stats{operation="ost_statfs",units="usecs"} 208
lustre_oss_ost_create_stats{operation="req_active",units="reqs"} 208
lustre_oss_ost_create_stats{operation="req_qdepth",units="reqs"} 208
lustre_oss_ost_create_stats{operation="req_timeout",units="secs"} 208
lustre_oss_ost_create_stats{operation="req_waittime",units="usecs"} 208
lustre_oss_ost_create_stats{operation="reqbuf_avail",units="bufs"} 428
# HELP lustre_stats_total Number of operations the filesystem has performed.
# TYPE lustre_stats_total counter
lustre_stats_total{component="mdt",operation="statfs",target="fs-MDT0000"} 133
# HELP lustre_stats_time_min Minimum time taken for an operation in microseconds.
# TYPE lustre_stats_time_min gauge
lustre_stats_time_min{component="mdt",operation="statfs",target="fs-MDT0000"} 0
# HELP lustre_stats_time_max Maximum time taken for an operation in microseconds.
# TYPE lustre_stats_time_max gauge
lustre_stats_time_max{component="mdt",operation="statfs",target="fs-MDT0000"} 57
# HELP lustre_stats_time_total Total time taken for an operation in microseconds.
# TYPE lustre_stats_time_total gauge
lustre_stats_time_total{component="mdt",operation="statfs",target="fs-MDT0000"} 1609
# EOF
